### ISSUE TITLE
An error occurred while calling the create inference API in the ES cluster upgraded from version 8.9 to 8.15.

### ISSUE THREAD

**Issue by Huaixinww** [elastic/elasticsearch#112694]
### Elasticsearch Version

8.9

### Java Version

_bundled_

### OS Version

x86_64 GNU/Linux

### Problem Description

After upgrading our ES cluster from version 8.9 to 8.15, we encountered the following error when calling the create inference API:
```
{
  "error": {
    "root_cause": [
      {
        "type": "status_exception",
        "reason": "Failed to store inference endpoint [os-embeddings]"
      }
    ],
    "type": "status_exception",
    "reason": "Failed to store inference endpoint [os-embeddings]",
    "caused_by": {
      "type": "null_pointer_exception",
      "reason": null
    }
  },
  "status": 500
}
```
We noticed that in the 8.15 cluster, the `mappings_versions` field in the `nodes_versions` parameter of `GET _cluster/state` contains information about relevant system indices. 
```
"nodes_versions": [
    {
      "node_id": "XpcbHlUSQR6Nq3iQ9ZmBHA",
      "transport_version": "8702003",
      "mappings_versions": {
        ".triggered_watches": {
          "version": 1,
          "hash": -502826165
        },
        ".secrets-inference": {
          "version": 1,
          "hash": -1434574148
        },
        ...  # other system index mappings_versions
      }
    },
```

However, in the cluster upgraded from 8.9, this `mappings_versions` field is empty. 
```
"nodes_versions": [
    {
      "node_id": "7SHcBQoRSSOBA2KXyz42hQ",
      "transport_version": "8702003",
      "mappings_versions": {}
    },
    {
      "node_id": "KFVdkkGDQ5qx-SoxbK8zxg",
      "transport_version": "8702003",
      "mappings_versions": {}
    },
    {
      "node_id": "RpOqtyQJRBuxHpq78T_8Gw",
      "transport_version": "8702003",
      "mappings_versions": {}
    }
  ]
```

This causes a NPE when calling the inference API, as shown in the following error trace:
```
{
  "error": {
    "root_cause": [
      {
        "type": "status_exception",
        "reason": "Failed to store inference endpoint [os-embeddings]",
        "stack_trace": """org.elasticsearch.ElasticsearchStatusException: Failed to store inference endpoint [os-embeddings]
	at org.elasticsearch.xpack.inference.registry.ModelRegistry.lambda$getStoreModelListener$5(ModelRegistry.java:334)
	at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:249)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:32)
	at org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:202)
	at org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:196)
	at org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onResponse(ActionListenerImplementations.java:307)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:32)
	at org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:95)
	at org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onResponse(ActionListenerImplementations.java:307)
	at org.elasticsearch.action.bulk.BulkOperation.completeBulkOperation(BulkOperation.java:396)
	at org.elasticsearch.action.bulk.BulkOperation.redirectFailuresOrCompleteBulkOperation(BulkOperation.java:391)
	at org.elasticsearch.action.bulk.BulkOperation.executeBulkRequestsByShard(BulkOperation.java:356)
	at org.elasticsearch.action.bulk.BulkOperation.doRun(BulkOperation.java:169)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.action.bulk.TransportBulkAction.executeBulk(TransportBulkAction.java:539)
	at org.elasticsearch.action.bulk.TransportBulkAction$1.doRun(TransportBulkAction.java:324)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:984)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.NullPointerException
	at java.util.Objects.requireNonNull(Objects.java:233)
	at org.elasticsearch.indices.SystemIndexDescriptor.getDescriptorCompatibleWith(SystemIndexDescriptor.java:622)
	at org.elasticsearch.action.admin.indices.create.AutoCreateAction$TransportAction$CreateIndexTask.execute(AutoCreateAction.java:307)
	at org.elasticsearch.action.admin.indices.create.AutoCreateAction$TransportAction.lambda$new$0(AutoCreateAction.java:120)
	at org.elasticsearch.cluster.service.MasterService.innerExecuteTasks(MasterService.java:1078)
	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:1041)
	at org.elasticsearch.cluster.service.MasterService.executeAndPublishBatch(MasterService.java:240)
	at org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.lambda$run$2(MasterService.java:1694)
	at org.elasticsearch.action.ActionListener.run(ActionListener.java:444)
	at org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.run(MasterService.java:1691)
	at org.elasticsearch.cluster.service.MasterService$5.lambda$doRun$0(MasterService.java:1286)
	at org.elasticsearch.action.ActionListener.run(ActionListener.java:444)
	at org.elasticsearch.cluster.service.MasterService$5.doRun(MasterService.java:1265)
	... 5 more
"""
      }
    ],
    "type": "status_exception",
    "reason": "Failed to store inference endpoint [os-embeddings]",
    "caused_by": {
      "type": "null_pointer_exception",
      "reason": null,
      "stack_trace": """java.lang.NullPointerException
	at java.util.Objects.requireNonNull(Objects.java:233)
	at org.elasticsearch.indices.SystemIndexDescriptor.getDescriptorCompatibleWith(SystemIndexDescriptor.java:622)
	at org.elasticsearch.action.admin.indices.create.AutoCreateAction$TransportAction$CreateIndexTask.execute(AutoCreateAction.java:307)
	at org.elasticsearch.action.admin.indices.create.AutoCreateAction$TransportAction.lambda$new$0(AutoCreateAction.java:120)
	at org.elasticsearch.cluster.service.MasterService.innerExecuteTasks(MasterService.java:1078)
	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:1041)
	at org.elasticsearch.cluster.service.MasterService.executeAndPublishBatch(MasterService.java:240)
	at org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.lambda$run$2(MasterService.java:1694)
	at org.elasticsearch.action.ActionListener.run(ActionListener.java:444)
	at org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.run(MasterService.java:1691)
	at org.elasticsearch.cluster.service.MasterService$5.lambda$doRun$0(MasterService.java:1286)
	at org.elasticsearch.action.ActionListener.run(ActionListener.java:444)
	at org.elasticsearch.cluster.service.MasterService$5.doRun(MasterService.java:1265)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:984)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.lang.Thread.run(Thread.java:1583)
"""
    },
    "stack_trace": """org.elasticsearch.ElasticsearchStatusException: Failed to store inference endpoint [os-embeddings]
	at org.elasticsearch.xpack.inference.registry.ModelRegistry.lambda$getStoreModelListener$5(ModelRegistry.java:334)
	at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:249)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:32)
	at org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:202)
	at org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:196)
	at org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onResponse(ActionListenerImplementations.java:307)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:32)
	at org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:95)
	at org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onResponse(ActionListenerImplementations.java:307)
	at org.elasticsearch.action.bulk.BulkOperation.completeBulkOperation(BulkOperation.java:396)
	at org.elasticsearch.action.bulk.BulkOperation.redirectFailuresOrCompleteBulkOperation(BulkOperation.java:391)
	at org.elasticsearch.action.bulk.BulkOperation.executeBulkRequestsByShard(BulkOperation.java:356)
	at org.elasticsearch.action.bulk.BulkOperation.doRun(BulkOperation.java:169)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.action.bulk.TransportBulkAction.executeBulk(TransportBulkAction.java:539)
	at org.elasticsearch.action.bulk.TransportBulkAction$1.doRun(TransportBulkAction.java:324)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:984)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.NullPointerException
	at java.util.Objects.requireNonNull(Objects.java:233)
	at org.elasticsearch.indices.SystemIndexDescriptor.getDescriptorCompatibleWith(SystemIndexDescriptor.java:622)
	at org.elasticsearch.action.admin.indices.create.AutoCreateAction$TransportAction$CreateIndexTask.execute(AutoCreateAction.java:307)
	at org.elasticsearch.action.admin.indices.create.AutoCreateAction$TransportAction.lambda$new$0(AutoCreateAction.java:120)
	at org.elasticsearch.cluster.service.MasterService.innerExecuteTasks(MasterService.java:1078)
	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:1041)
	at org.elasticsearch.cluster.service.MasterService.executeAndPublishBatch(MasterService.java:240)
	at org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.lambda$run$2(MasterService.java:1694)
	at org.elasticsearch.action.ActionListener.run(ActionListener.java:444)
	at org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.run(MasterService.java:1691)
	at org.elasticsearch.cluster.service.MasterService$5.lambda$doRun$0(MasterService.java:1286)
	at org.elasticsearch.action.ActionListener.run(ActionListener.java:444)
	at org.elasticsearch.cluster.service.MasterService$5.doRun(MasterService.java:1265)
	... 5 more
"""
  },
  "status": 500
}
```

### Steps to Reproduce

1. An ES cluster upgraded from version 8.9 to 8.15.
2. Calling the create inference API interface.

**Comment by weizijun** [comment]
> This is a system indices issue, not ML

hi, @thecoop, this bug will result in mappings_versions map empty, It will lead creating system indices error. It's a serious problem.  Is there any progress now?

**Comment by ldematte** [comment]
After a first investigation, I think this is the specular of https://github.com/elastic/elasticsearch/issues/112946: in a mixed cluster situation, if a new (e.g. 8.15) node comes up, it can trigger the creation of a inference system indices. But that will be carried on by the master node; if the master is still on an older version (e.g. 8.9) its `SystemIndices` will not have any system index descriptor for the inference indices, and that's where the NPE comes from. 
I'll keep investigating to understand where/what/when the creation is triggered, so we can decide how to best prevent a "too early" creation of the indices.

**Comment by ldematte** [comment]
@davidkyle where you able to repro the issue?

**Comment by davidkyle** [comment]
> @davidkyle where you able to repro the issue?

No. I wrote an upgrade test that upgraded from 8.9 to 8.15 then in the fully upgraded cluster the test creates an inference end point causing the creation of the system index. It did not fail in 10 runs.

**Comment by ldematte** [comment]
Same here, I was not able to reproduce it in a fully upgraded cluster.

I'll try to reproduce it under very specific circumstances to test what I _think_ might trigger this behaviour:
- the cluster is _not_ fully upgraded
- the master node is still 8.9
- the inference API gets called on an upgraded node (8.9 does not have the endpoint)

**Comment by weizijun** [comment]
> Same here, I was not able to reproduce it in a fully upgraded cluster.
> 
> I'll try to reproduce it under very specific circumstances to test what I _think_ might trigger this behaviour:
> 
> * the cluster is _not_ fully upgraded
> * the master node is still 8.9
> * the inference API gets called on an upgraded node (8.9 does not have the endpoint)

This case is a rolling upgrade from 8.9 to 8.15. If all 8.9 nodes shutdown, it maybe not be reproducible.

**Comment by ldematte** [comment]
I was not able to reproduce even with a rolling upgrade, _after_ all the nodes got upgraded to 8.15.
I will try to reproduce it _during_ the upgrade, when we are in a mixed cluster scenario. Based on my understanding of the source code, I think this can happen only if we have a 8.9 master.

**Comment by weizijun** [comment]
> I was not able to reproduce even with a rolling upgrade, _after_ all the nodes got upgraded to 8.15.
> I will try to reproduce it _during_ the upgrade, when we are in a mixed cluster scenario. Based on my understanding of the source code, I think this can happen only if we have a 8.9 master.

If the `mappings_versions ` from command `GET _cluster/state` is empty, it will cause NPE.

**Comment by weizijun** [comment]
I guess the bug is from the PR https://github.com/elastic/elasticsearch/pull/99114 And  https://github.com/elastic/elasticsearch/pull/99555

**Comment by davidkyle** [comment]
@weizijun Did the upgraded cluster eventually resolve itself? By that I mean did the missing `mappings_versions` appear and were you able to create the Inference endpoint? We would like to know if the issue was transient or if the cluster ever recovered. Thank you

**Comment by weizijun** [comment]
> Did the upgraded cluster eventually resolve itself? By that I mean did the missing `mappings_versions` appear and were you able to create the Inference endpoint? We would like to know if the issue was transient or if the cluster ever recovered. Thank you

The cluster will never recovery. It can't create the Inference endpoint because of the NPE of creating inference index. We will provide the logs of master to see what happen.

**Comment by ldematte** [comment]
Just a head's up: I talked to @rjernst and the missing "mappings_versions" can be a red herring and unrelated to the NPE. 
Still, the rest stands: the descriptor in SystemIndices are declared statically (in code) by the plugin, and read during NodeConstruction, so if they are not present (which is why we have a NPE) something is probably going wrong on the master node. 
It's pretty important to have logs especially from the early phases (node start), as SystemIndices is created at start and never changes.

**Comment by weizijun** [comment]
Hi, @ldematte , I noticed that, our version upgrade process is to add all new version nodes to the cluster,  all data transfer from old version nodes to new version nodes, and then shutdown all old nodes. I send the master log of new node to @davidkyle.

**Comment by ldematte** [comment]
Hello @weizijun, I got the log file, thanks. 
Unfortunately it's very hard to work out what's going on without logs from all nodes tho (old and new), especially the master (from the logs, this is `es-cn-35t3xjo7a000190t2-data-i-1` initially, then `es-cn-35t3xjo7a000190t2-data-i-0` -- both ver 8.9 -- and then becomes `es-cn-35t3xjo7a000190t2-data-i2-2` -- ver 8.15.1).
One thing we need to understand is why the newly-started node (`es-cn-35t3xjo7a000190t2-data-i2-2`) becomes master straight away rather than joining an existing master.
Is your upgrade process doing something to force that? What else happens during your upgrade process? Allocation filtering, manual killing of nodes, or anything else?

**Comment by ldematte** [comment]
Without complete logs for the whole cluster and a detailed description of what is happening/the actions performed on the cluster I'm afraid there is very little we can do.

**Comment by ldematte** [comment]
> The cluster will never recovery. It can't create the Inference endpoint because of the NPE of creating inference index. We will provide the logs of master to see what happen.

Have you tried restarting the nodes, after the upgrade, one by one? Using the API, not killing and restarting.

**Comment by weizijun** [comment]
step 1:
node1: v8.9
elasticsearch.yml:
```
node.name: node-1
cluster.initial_master_nodes: ["node-1"]
```

start node1

step 2:
node2: v8.15
elasticsearch.yml:
```
node.name: node-2
cluster.initial_master_nodes: ["node-1", "node-2"]
```

start node2

step 3:
node3: v8.15
elasticsearch.yml:
```
node.name: node-3
cluster.initial_master_nodes: ["node-1", "node-2", "node-3"]
```

start node3

step 4:
then stop node1

step 5:
then this api will returen NPE:
```
PUT _inference/sparse_embedding/my-elser-model
{
  "service": "elser",
  "service_settings": {
    "num_allocations": 1,
    "num_threads": 1
  },
  "task_settings": {}
}

return :
{
  "error" : {
    "root_cause" : [
      {
        "type" : "status_exception",
        "reason" : "Failed to store inference model [my-elser-model]"
      }
    ],
    "type" : "status_exception",
    "reason" : "Failed to store inference model [my-elser-model]",
    "caused_by" : {
      "type" : "null_pointer_exception",
      "reason" : null
    }
  },
  "status" : 500
}
```

**Comment by ldematte** [comment]
I was able to reproduce it with the sequence described. 
Looks like we are missing a fixup: when the last 8.9 node leaves the cluster, the system indices descriptors/their mappings versions are not update to reflect what is now available to the cluster.
Restarting the master will force the update, but that should not be needed.

**Comment by ldematte** [comment]
The issue has been fixed for 8.17 (https://github.com/elastic/elasticsearch/pull/115771) and is being backported to 8.16.1 too (https://github.com/elastic/elasticsearch/pull/116646)
